### ISSUE TITLE
[TNSRE 투고] 채널별 Uncertainty 방법론 노벨티 강화 전략 및 증거 정리

### ISSUE THREAD

**Issue by kimshhhhh** [kimshhhhh/eeg_uncertainty#2]
## 개요

본 이슈는 **채널별 EEG Uncertainty Estimation** 연구를 TNSRE(IEEE Transactions on Neural Systems and Rehabilitation Engineering)에 투고하기 위한 방법론 노벨티 강화 전략과 현재 코드베이스에서 도출된 증거를 정리합니다.

---

### 핵심 Uncertainty 공식

$$U(c, t) = \alpha \cdot \text{LocalNonStat}(c, t) + \beta \cdot (1 - \text{InterCorr}(c, t)) \cdot (1 - \text{Relevance}(c))$$

| 구성 요소 | 구현 파일 | 의미 |
|---------|---------|------|
| `LocalNonStat(c, t)` | `channel_wise_method_3.py:compute_local_nonstationarity()` | 슬라이딩 윈도우 분산으로 시간-국소 비정상성 측정 |
| `InterCorr(c, t)` | `channel_wise_method_3.py:compute_inter_channel_correlation()` | 인접 전극 그래프 기반 공간 상관관계 |
| `Relevance(c)` | `channel_wise_method_3.py:compute_task_relevance()` | Mutual Information으로 과제-채널 관련도 측정 |

### 모델 통합 방식

- **Method 4** (`eegnetwithuncertainty_4.py`): Spatial Attention Gate — `gate = exp(-β * û)` 로 고-uncertainty 채널 신호를 soft-suppress
- **Method 5** (`eegnetwithuncertainty_5.py`): Progressive Channel Masking — 학습이 진행될수록 high-uncertainty 채널을 점진적으로 제거 (keep_ratio: 1.0 → 0.5)

---

### 1. 세 요소 결합의 이론적 정당화 (핵심)

**현황 문제**: 현재 `alpha * NonStat + beta * (1-Corr) * (1-Rel)` 공식이 휴리스틱으로 보일 수 있음.

**강화 방향**: 각 구성 요소를 정보이론으로 연결
- `LocalNonStat` → **Differential Entropy 상한** : 비정상 구간에서 채널 신호의 정보 불확실성 증가
- `(1-InterCorr)` → **조건부 독립성 이탈** : 인접 채널과의 상관도가 낮을수록 해당 채널이 EEG 신호 공간에서 outlier
- `(1-Relevance)` → **Task-Irrelevant 신호를 noise로 취급** : MI가 낮은 채널의 비정상성은 motor imagery와 무관한 artifact일 확률 높음

**추가 근거**: BCIC2b (3채널)에서 동일 공식이 `build_adjacency()` 자동 적응 (`channel_wise_method_3_savenpz_bcic2b.py:build_adjacency()`) → 채널 수에 무관한 범용성 입증 가능

---

### 2. 신호-기반 vs 모델-기반 Uncertainty 대비

**현황**: MC-Dropout 기반 trial-level uncertainty (`training_mcdropout_channel_wise.py`) 와 신호-기반 채널 uncertainty가 병존하지만 비교 실험 없음.

**강화 방향**: 두 방식의 *일치도* 를 증거로 제시

| 비교 항목 | 신호-기반 (proposed) | MC-Dropout (baseline) |
|---------|------|------|
| 해상도 | 채널 × 시간 (C×T) | 트라이얼 레벨 (scalar) |
| 계산 비용 | O(C·T) | O(T·forward_passes) |
| 해석 가능성 | 물리적 (분산/상관) | 블랙박스 |
| EEG artifact 민감도 | ✅ 직접 반영 | ❌ 간접적 |

**증거 확보 방법**: `results/250630_model_uncertainty_deepconvnet_all_eval/sub_*.npz` (MC-Dropout trial uncertainty) vs `results/260309_uncertainty_npz_bcic2b/` (signal-based channel uncertainty) 의 Spearman 상관 분석

---

### 3. 공간 그래프 구조의 신경생리학적 근거 강화

**현황**: 인접 전극 그래프는 물리적 거리 기반 (`adjacency` dict in `channel_wise_method_3.py:17-42`)

**강화 방향**: 다음 두 가지로 노벨티 보강
1. **Neurophysiology 정렬 검증**: C3(ch7), Cz(ch9), C4(ch11) — motor cortex 채널의 uncertainty가 과제 수행 중 *낮아야* 함 (Motor Imagery에서 event-related desynchronization이 명확하므로). 현재 `results/topo/` 의 토폴로지 맵 활용
2. **Graph 구조 Ablation**: 인접 그래프 vs 전체 연결 vs 무작위 연결 → 공간 그래프의 효과 입증

---

### 4. Cross-Dataset 일반화 (BCIC2a → BCIC2b)

**현황**: BCIC2a (22채널, 4-class), BCIC2b (3채널, 2-class) 모두 구현 완료

**강화 방향**: 노벨티 포인트 — *채널 수와 task에 독립적인 uncertainty 공식*
- BCIC2a: 22채널, Fz/FC/C/CP/P 전극 배치, 인접 그래프 22노드
- BCIC2b: 3채널 (C3/Cz/C4), `build_adjacency(3)` 자동 생성
- 두 데이터셋 모두 동일 공식 적용 → 방법론의 *scalability* 가 TNSRE novelty claim 핵심

**필요 실험**: 각 dataset에서 uncertainty-guided model (Method 4/5) vs baseline EEGNet 정확도 비교

---

### 5. Uncertainty-Guided Training의 노벨티 포인트

**Method 4** (`eegnetwithuncertainty_4.py:67-80`):
```python
u_hat = self.spatial_attn(x_feat)   # [B, 1, C, T]
gate = torch.exp(-self.beta * u_hat) # exponential soft-mask
x_feat = x_feat * gate
```
→ 사전 계산된 신호 기반 uncertainty `u`로 `u_hat`을 지도학습 → **Signal Prior로부터 Attention을 학습**하는 새로운 패러다임

**Method 5** (`eegnetwithuncertainty_5.py:59-70`):
```python
keep_ratio = 1.0 - 0.5 * progress  # curriculum: 점진적 채널 제거
k = int(C * keep_ratio)
_, idx = torch.topk(-u_mean[b], k)  # low-uncertainty 채널만 유지
```
→ Curriculum Learning으로 고-uncertainty 채널을 점진적 제거 → **Uncertainty-driven Curriculum Learning** 으로 포지셔닝 가능

---

### E1. Uncertainty Map vs ERSP 정렬
- **증거 파일**: `results/250430_channel_wise_2/sub{s}_trial{t}_ch{c}.png`
- **내용**: 각 채널별 uncertainty curve + Alpha/Beta band ERSP 상위/하위 15% 구간 overlay
- **TNSRE 활용**: uncertainty 피크가 비-ERSP 구간 (artifact, baseline noise)과 일치함을 보이면 neurophysiology 근거 확보

### E2. 신호 기반 Uncertainty의 Trial-Level 일관성
- **증거 파일**: `results/260309_uncertainty_npz_bcic2b/sub*/trial_*_uncertainty.npz`
- **저장 keys**: `uncertainty` (C×T), `local_variance` (C×T), `inter_channel_correlation` (C×T), `relevance_scores` (C,), `class_label`
- **활용**: 클래스별 uncertainty 패턴 비교 → left hand vs right hand에서 C3/C4 uncertainty 차이

### E3. Curriculum Masking의 학습 효과
- **증거**: `training_with_uncertainty_eegnet.py` — uncertainty vector로 augmentation 후 학습
- **비교 대상**: `training_baseline.py` (baseline EEGNet, no uncertainty)
- **필요 실험**: Method 5의 점진적 채널 제거가 수렴 속도/최종 정확도에 미치는 영향

### E4. 모델 Uncertainty와의 Correlation
- **증거 파일**: `results/250630_model_uncertainty_deepconvnet_all_eval/sub_*.npz` (keys: `uncertainties`, `labels`)
- **분석 코드**: `uncertainty.py` — point-biserial correlation(Unc, Wrong) 이미 구현
- **활용**: 신호-기반 채널 uncertainty의 평균값과 MC-Dropout trial uncertainty 간 Spearman ρ 계산

---

## TNSRE 투고를 위한 추가 실험 체크리스트

- [ ] **E1 정량화**: ERSP 상위 구간 vs uncertainty peak overlap ratio (F1-score)
- [ ] **E2 분석**: 클래스별 C3/Cz/C4 채널 uncertainty 평균 (left hand vs right hand)
- [ ] **E3 실험**: Method 4 vs Method 5 vs Baseline accuracy on BCI IV-2a (9 subjects)
- [ ] **E4 분석**: Signal-based uncertainty mean vs MC-Dropout trial uncertainty — Spearman ρ per subject
- [ ] **Ablation**: `alpha=1,beta=0` (NonStat only) vs `alpha=0,beta=1` (Corr×Rel only) vs `alpha=0.5,beta=0.5`
- [ ] **Cross-dataset**: BCIC2b에서 동일 pipeline 정확도 보고
- [ ] **Calibration**: ECE (Expected Calibration Error) 평가 — uncertainty가 실제 오류율과 얼마나 calibrated

---

## Related Work 포지셔닝

| 선행연구 | 채널별 Uncertainty | 신호 기반 | 모델 통합 | 비교 대상 |
|---------|:-----------------:|:--------:|:--------:|:--------:|
| Milanés-Hermosilla et al. 2021 (MC-Dropout BCI) | ❌ trial-level | ❌ model | ❌ | BCIC2a 77.43% |
| MCD-EEGNet 2025 (MDPI) | ❌ LayerCAM | ❌ model | ✅ | private dataset |
| Wang et al. CVPR 2023 (gradient attr) | ✅ | ❌ | ❌ | vision only |
| **Proposed** | ✅ **C×T resolution** | ✅ **signal prior** | ✅ **attn/curriculum** | BCIC2a + BCIC2b |

**핵심 차별점**: 기존 연구 중 (1) 채널별 시공간 해상도의 uncertainty를 (2) 신호 통계로부터 직접 유도하고 (3) 이를 attention 학습의 supervision signal로 사용한 연구 없음.

---

## 논문 구조 제안 (TNSRE 형식 기준)

1. **Introduction**: EEG nonstationarity 문제 → 채널별 uncertainty의 필요성
2. **Related Work**: MC-Dropout BCI / Channel Attention / Uncertainty Attribution
3. **Method**:
   - 3.1 Signal-based Channel Uncertainty (공식 유도 + 이론적 정당화)
   - 3.2 Uncertainty-Guided Spatial Attention (Method 4)
   - 3.3 Uncertainty-driven Curriculum Learning (Method 5)
4. **Experiments**: BCIC2a (4-class, 22ch), BCIC2b (2-class, 3ch)
5. **Analysis**: ERSP 정렬, 신경생리학적 해석, Ablation, Calibration
6. **Conclusion**

---

*Generated by oh-my-agents — 2026-04-08*
